File: software_architecture_model_python/src/software_architecture_model_python/validators/classifier.py
"""
Ταξινόμηση αρχιτεκτονικής project με βάση canonical trees και signatures.

Το module παρέχει τον ArchitectureClassifier, ο οποίος συγκρίνει το input tree
με όλες τις διαθέσιμες αρχιτεκτονικές και επιστρέφει score ομοιότητας.
"""

from software_architecture_model_python.schemas.registry import SchemaRegistry
from software_architecture_model_python.tree.comparator import TreeComparator
from software_architecture_model_python.tree.scanner import TreeScanner
from software_architecture_model_python.validators.signature import (
    ArchitectureSignature,
    SignatureEngine,
)


class ArchitectureClassifier:
    """
    Κάνει ταξινόμηση αρχιτεκτονικής:
    - MVC
    - MVVM
    - DDD
    - Event-Driven
    - Flow-Based Architecture
    """

    def __init__(self) -> None:
        self.registry = SchemaRegistry()
        self.scanner = TreeScanner()
        self.comparator = TreeComparator()
        self.signatures = SignatureEngine().load_signatures()

    def classify(self, project_path: str) -> dict[str, float]:
        """
    Υπολογίζει score ομοιότητας για κάθε αρχιτεκτονική.

    :param project_path: Το μονοπάτι του project.
    :type project_path: str
    :return: Λεξικό αρχιτεκτονική → score.
    :rtype: dict[str, float]
    """
        input_tree = self.scanner.scan(project_path)
        results: dict[str, float] = {}

        # Extract input nodes once
        input_nodes = set()
        for parent, children in input_tree.items():
            input_nodes.add(parent)
            input_nodes.update(children)

        for arch_name, canonical_tree in self.registry.load_all().items():
            # 1) Jaccard score (TreeComparator)
            comparison = self.comparator.compare(input_tree, canonical_tree)
            jaccard_score = comparison["jaccard"]

            # 2) Signature score (SignatureEngine)
            signature = self.signatures[arch_name]
            sig_score = self.signature_score(input_nodes, signature)

            # 3) Combine scores
            final_score = (jaccard_score * 0.6) + (sig_score * 0.4)

            # 4) Save result
            results[arch_name] = round(final_score, 3)

        return results

    def best_match(self, project_path: str) -> tuple[str, float]:
        """
    Επιστρέφει την αρχιτεκτονική με το υψηλότερο score.

    :param project_path: Το μονοπάτι του project.
    :type project_path: str
    :return: Ζεύγος (όνομα αρχιτεκτονικής, score).
    :rtype: tuple[str, float]
    """
        scores = self.classify(project_path)
        best = max(scores.items(), key=lambda x: x[1])
        return best

    def signature_score(self, input_nodes: set[str], signature: ArchitectureSignature) -> float:
        """
    Υπολογίζει score ομοιότητας με βάση την αρχιτεκτονική υπογραφή (signature).

    :param input_nodes: Οι κόμβοι του input tree.
    :type input_nodes: set[str]
    :param signature: Η υπογραφή της αρχιτεκτονικής.
    :type signature: ArchitectureSignature
    :return: Score ομοιότητας (0–1).
    :rtype: float
    """
        required_match = len(input_nodes & signature.required_nodes) / max(
            1, len(signature.required_nodes)
        )
        optional_match = len(input_nodes & signature.optional_nodes) / max(
            1, len(signature.optional_nodes)
        )
        return round((required_match * 0.7) + (optional_match * 0.3), 3)

File: js_package/src/validators/classifier.ts
/**
 * Ταξινόμηση αρχιτεκτονικής project με βάση canonical trees και signatures.
 *
 * Ο ArchitectureClassifier:
 * - συγκρίνει το input tree με όλες τις διαθέσιμες αρχιτεκτονικές
 * - υπολογίζει Jaccard similarity (TreeComparator)
 * - υπολογίζει Signature similarity (SignatureEngine)
 * - επιστρέφει score ομοιότητας για κάθε αρχιτεκτονική
 */

import { SchemaRegistry } from "../schemas/registry.js";
import { TreeScanner } from "../tree/scanner.js";
import { TreeComparator } from "../tree/comparator.js";
import { SignatureEngine, ArchitectureSignature } from "../validators/signature.js";
import { intersection } from "../utils/set.js";


export class ArchitectureClassifier {
  private registry: SchemaRegistry;
  private scanner: TreeScanner;
  private comparator: TreeComparator;
  private signatures: Record<string, ArchitectureSignature>;

  constructor() {
    this.registry = new SchemaRegistry();
    this.scanner = new TreeScanner();
    this.comparator = new TreeComparator();
    this.signatures = new SignatureEngine().load_signatures();
  }

  /**
   * Υπολογίζει score ομοιότητας για κάθε αρχιτεκτονική.
   */
  async classify(projectPath: string): Promise<Record<string, number>> {
    const inputTree = await this.scanner.scan(projectPath);
    const results: Record<string, number> = {};

    // Extract input nodes once
    const inputNodes = new Set<string>();
    for (const parent of Object.keys(inputTree)) {
      inputNodes.add(parent);
      for (const child of inputTree[parent]) {
        inputNodes.add(child);
      }
    }

    const canonical = await this.registry.loadAll();

    for (const archName of Object.keys(canonical)) {
      // 1) Jaccard score (TreeComparator)
      const comparison = this.comparator.compare(inputTree, canonical[archName]);
      const jaccardScore = comparison.jaccard;

      // 2) Signature score (SignatureEngine)
      const signature = this.signatures[archName];
      const sigScore = this.signature_score(inputNodes, signature);

      // 3) Combine scores
      const finalScore = (jaccardScore * 0.6) + (sigScore * 0.4);

      // 4) Save result
      results[archName] = Number(finalScore.toFixed(3));
    }

    return results;
  }

  /**
   * Επιστρέφει την αρχιτεκτονική με το υψηλότερο score.
   */
  async best_match(projectPath: string): Promise<[string, number]> {
    const scores = await this.classify(projectPath);
    let bestArch = "";
    let bestScore = -1;

    for (const [arch, score] of Object.entries(scores)) {
      if (score > bestScore) {
        bestArch = arch;
        bestScore = score;
      }
    }

    return [bestArch, bestScore];
  }

  /**
   * Υπολογίζει score ομοιότητας με βάση την αρχιτεκτονική υπογραφή (signature).
   */
  private signature_score(
    inputNodes: Set<string>,
    signature: ArchitectureSignature
  ): number {
    const requiredMatch =
      intersection(inputNodes, signature.required_nodes).size /
      Math.max(1, signature.required_nodes.size);

    const optionalMatch =
      intersection(inputNodes, signature.optional_nodes).size /
      Math.max(1, signature.optional_nodes.size);

    return Number(((requiredMatch * 0.7) + (optionalMatch * 0.3)).toFixed(3));
  }
}
